core-bundle/src/Twig/Extension/ContaoTemplateExtension.php

```diff
@@ -76,4 +91,18 @@ public function renderContaoBackendTemplate(array $blocks = []): string
 
         return $response->getContent();
     }
+
+    /**
+     * THis filter is a drop-in replacement for 'raw' but can be substituted
+     * with a harmless identity function (i.e. removing 'is_safe') once we're
+     * switching to output encoding.
+     */
+    public function handleInputEncoding(array $context, string $input): string
+    {
+        if (($context['_contao_encoding'] ?? null) !== 'input') {
+            throw new \LogicException("The filter 'c' must only be applied to Contao template context. Did you mean 'raw'?");
+        }
+
+        return $input;
```

Comment:
How about using

```suggestion
        return htmlspecialchars($input, ENT_QUOTES, null, false);
```

With `double_encode` set to `false`. This should keep all “input encoded” strings the same and would force you to use `|raw` if the actual raw value needs to be written to the output.

This would reduce the risk of template developers learing the wrong way to work with twig templates. The only difference to “regular” twig templates would be to use `|c` everywhere. Maybe we could make the `c` filter the default ouput filter for Contao twig templates?

Reply:
Imho this is the wrong way round - we need a filter that allows passing "unsafe" HTML because it's already pre-rendered and sanitized. With `htmlspecialchars` your would get the same as defining the filter without `'is_safe' => 'html`. 

See https://twig.symfony.com/doc/3.x/advanced.html#automatic-escaping

> Maybe we could make the c filter the default ouput filter for Contao twig templates?

This should be possible on a template level with something like the following if we'd register an `Escaper`:

```twig
{% autoescape 'contao' %}
<h2>My Template</h2>
Some {{ content }} that is treated like being passed through the raw filter.
{% endautoescape %}
```

I think `is_safe` can also be evaluated by a callback (context sensitive), so we could maybe check the context before blindly allowing anything.

Reply:
Ah! I just found out that you can mark classes as safe for the escaper via `EscaperExtension::addSafeClass():` - I'll investigate if this could be the solution. Maybe we can simply proxy everything through this class when building the context from within a Contao template. :smile:

Reply:
Yes, this actually works!! :tada: :tada:
I'll finish this tomorrow.

Reply:
Done :star: